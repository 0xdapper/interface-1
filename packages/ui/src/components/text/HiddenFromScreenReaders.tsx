import { PropsWithChildren } from 'react'
import { View } from 'react-native'

export function HiddenFromScreenReaders<T>({ children }: PropsWithChildren<T>): JSX.Element {
  // TODO(MOB-1533) Make hidden from screen reader functionality work with web too
  return (
    <View accessibilityElementsHidden={true} importantForAccessibility="no-hide-descendants">
      {children}
    </View>
  )
}
