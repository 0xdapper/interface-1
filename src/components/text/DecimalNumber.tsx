import React from 'react'
import { Text, TextProps } from 'src/components/Text'
import { Theme } from 'src/styles/theme'

type DecimalNumberProps = TextProps & {
  number: string
  separator?: string
  variant: keyof Theme['textVariants']
}

// Utility component to display decimal numbers where the decimal part
// is dimmed
export function DecimalNumber({ number, separator = '.', variant, ...rest }: DecimalNumberProps) {
  const [pre, post] = number.split(separator)

  return (
    <Text variant={variant} {...rest}>
      {pre}
      {post && (
        <Text color="textTertiary" variant={variant}>
          {separator}
          {post}
        </Text>
      )}
    </Text>
  )
}
