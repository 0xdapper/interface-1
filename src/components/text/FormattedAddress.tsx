import React from 'react'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { shortenAddress } from 'src/utils/addresses'

type FormattedAddressProps = {
  name?: string | null
  address: string
} & { variant?: keyof Theme['textVariants'] }

export function FormattedAddress({ name, address, variant = 'body1' }: FormattedAddressProps) {
  return <Text variant={variant}>{name ?? shortenAddress(address)}</Text>
}
