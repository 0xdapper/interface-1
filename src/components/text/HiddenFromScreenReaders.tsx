import React, { PropsWithChildren } from 'react'
import { View } from 'react-native'

export function HiddenFromScreenReaders<T>({ children }: PropsWithChildren<T>) {
  return (
    <View accessibilityElementsHidden={true} importantForAccessibility="no-hide-descendants">
      {children}
    </View>
  )
}
