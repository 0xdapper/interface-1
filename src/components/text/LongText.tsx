/* eslint-disable react-native/no-inline-styles */
import React, { ComponentProps, useCallback, useReducer, useState } from 'react'
import { LayoutChangeEvent, NativeSyntheticEvent, TextLayoutEventData } from 'react-native'
import Markdown from 'react-native-markdown-display'
import { useAppTheme } from 'src/app/hooks'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { openUri } from 'src/utils/linking'

type LongTextProps = {
  initialDisplayedLines?: number
  text: string
  gap?: keyof Theme['spacing']
  color?: string
  linkColor?: string
  readMoreOrLessColor?: string
  renderAsMarkdown?: boolean
  variant?: keyof Theme['textVariants']
} & Omit<
  ComponentProps<typeof Text>,
  'children' | 'numberOfLines' | 'onTextLayout' | 'color' | 'variant'
>

export function LongText(props: LongTextProps): JSX.Element {
  const theme = useAppTheme()
  const {
    initialDisplayedLines = 3,
    text,
    gap = 'spacing8',
    color = theme.colors.textPrimary,
    linkColor = theme.colors.accentAction,
    readMoreOrLessColor = theme.colors.accentAction,
    renderAsMarkdown = false,
    variant = 'bodySmall',
    ...rest
  } = props
  const [maximized, toggleMaximized] = useReducer(
    (isMaximized) => !isMaximized,
    renderAsMarkdown ? true : false
  )
  const [textLengthExceedsLimit, setTextLengthExceedsLimit] = useState(false)
  const [initialContentHeight, setInitialContentHeight] = useState<number | undefined>(undefined)

  const textLineHeight = theme.textVariants[variant].lineHeight
  const maxVisibleHeight = textLineHeight * initialDisplayedLines

  const onLayout = useCallback(
    (event: LayoutChangeEvent) => {
      if (!renderAsMarkdown || initialContentHeight !== undefined) return
      const textContentHeight = event.nativeEvent.layout.height
      setTextLengthExceedsLimit(textContentHeight > maxVisibleHeight)
      toggleMaximized()
      setInitialContentHeight(textContentHeight)
    },
    [initialContentHeight, maxVisibleHeight, renderAsMarkdown]
  )

  const onTextLayout = useCallback(
    (e: NativeSyntheticEvent<TextLayoutEventData>) => {
      if (!renderAsMarkdown) {
        setTextLengthExceedsLimit(e.nativeEvent.lines.length >= initialDisplayedLines)
      }
    },
    [renderAsMarkdown, initialDisplayedLines]
  )

  return (
    <Flex gap={gap}>
      <Text
        numberOfLines={maximized ? undefined : initialDisplayedLines}
        style={
          renderAsMarkdown
            ? {
                color,
                height:
                  !textLengthExceedsLimit || maximized
                    ? 'auto'
                    : maxVisibleHeight - theme.spacing.spacing2,
                overflow: 'hidden',
              }
            : { color }
        }
        variant={variant}
        onLayout={onLayout}
        onTextLayout={onTextLayout}
        {...rest}>
        {renderAsMarkdown ? (
          <Markdown
            style={{
              body: { color },
              link: { color: linkColor },
              paragraph: {
                marginBottom: 0,
                marginTop: 0,
                fontSize: theme.textVariants.bodySmall.fontSize,
                lineHeight: theme.textVariants.bodySmall.lineHeight,
              },
            }}
            onLinkPress={(url): false => {
              // add our own custom link handler since it has security checks that only open http/https links
              openUri(url)
              return false
            }}>
            {text}
          </Markdown>
        ) : (
          text
        )}
      </Text>

      {textLengthExceedsLimit ? (
        <Text
          style={{ color: readMoreOrLessColor }}
          testID="read-more-button"
          variant="buttonLabelSmall"
          onPress={toggleMaximized}>
          {maximized ? 'Read less' : 'Read more'}
        </Text>
      ) : null}
    </Flex>
  )
}
