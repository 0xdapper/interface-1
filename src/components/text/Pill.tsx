import React, { ReactNode } from 'react'
import { Flex, FlexProps } from 'src/components/layout'
import { Text } from 'src/components/Text'

type PillProps = {
  customBackgroundColor?: string
  customBorderColor?: string
  foregroundColor?: string
  icon?: ReactNode
  label?: ReactNode
} & FlexProps

export function Pill({
  customBackgroundColor,
  children,
  customBorderColor,
  borderRadius = 'full',
  foregroundColor,
  flexDirection = 'row',
  icon,
  label,
  px = 'sm',
  py = 'xs',
  ...rest
}: PillProps) {
  return (
    <Flex
      alignItems="center"
      backgroundColor="gray200"
      borderColor="none"
      borderRadius={borderRadius}
      borderWidth={1}
      flexDirection={flexDirection}
      gap="xs"
      justifyContent="center"
      px={px}
      py={py}
      style={{
        ...(customBackgroundColor ? { backgroundColor: customBackgroundColor } : {}),
        ...(customBorderColor ? { borderColor: customBorderColor } : {}),
      }}
      {...rest}>
      {icon ?? null}
      {label ? (
        <Text style={{ color: foregroundColor }} variant="bodyMd">
          {label}
        </Text>
      ) : null}
      {children}
    </Flex>
  )
}
