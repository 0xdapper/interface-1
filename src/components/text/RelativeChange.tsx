import React from 'react'
import { Text } from 'src/components/Text'

interface RelativeChangeProps {
  change?: number
}

export function RelativeChange({ change }: RelativeChangeProps) {
  const direction: 'up' | 'down' | undefined = change ? (change > 0 ? 'up' : 'down') : undefined
  return (
    <Text
      color={direction === 'up' ? 'green' : direction === 'down' ? 'red' : 'gray600'}
      variant="bodySm">
      {direction === 'up' ? '↑ ' : direction === 'down' ? '↓ ' : ''}
      {change ? `${change.toFixed(1)}%` : '-'}
    </Text>
  )
}
