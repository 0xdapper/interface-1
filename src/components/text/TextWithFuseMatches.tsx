import Fuse from 'fuse.js'
import React from 'react'
import { TextProps } from 'react-native'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { Flex } from '../layout'

interface TextWithFuseMatchesProps {
  text: string
  matches?: readonly Fuse.FuseResultMatch[]
  variant?: keyof Theme['textVariants']
  numberOfLines?: TextProps['numberOfLines']
}

export function TextWithFuseMatches({
  matches,
  text,
  variant = 'body',
  numberOfLines = 1,
}: TextWithFuseMatchesProps & TextProps) {
  if (!matches || matches.length === 0) {
    return (
      <Text color="textColor" numberOfLines={numberOfLines} variant={variant}>
        {text}
      </Text>
    )
  }

  const charIsMatch = new Set()
  for (const match of matches) {
    for (const index of match.indices) {
      for (let i = index[0]; i < index[1] + 1; i++) {
        charIsMatch.add(i)
      }
    }
  }

  // PERF: batch pieces?
  const pieces = []
  for (let i = 0; i < text.length; i++) {
    if (charIsMatch.has(i)) {
      pieces.push([text[i], true])
    } else {
      pieces.push([text[i], false])
    }
  }

  const elements = (
    <>
      {pieces.map((p, i) => {
        if (p[1])
          return (
            <Text key={`${i}${p[0]}`} color="textColor" variant={variant}>
              {p[0]}
            </Text>
          )
        else
          return (
            <Text key={`${i}${p[0]}`} color="gray400" variant={variant}>
              {p[0]}
            </Text>
          )
      })}
    </>
  )

  return <Flex row>{elements}</Flex>
}
