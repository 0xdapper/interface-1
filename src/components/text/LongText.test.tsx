import React from 'react'
import { LongText } from 'src/components/text/LongText'
import { renderWithTheme } from 'src/test/render'

it('renders a LongText', () => {
  const tree = renderWithTheme(<LongText text="Some very long text" />)

  expect(tree).toMatchSnapshot()
})
