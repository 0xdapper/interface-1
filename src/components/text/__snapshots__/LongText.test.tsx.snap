// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a LongText 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    numberOfLines={3}
    onLayout={[Function]}
    onTextLayout={[Function]}
    style={
      Array [
        Object {
          "color": "#0D111C",
          "display": "flex",
          "fontFamily": "Inter-Regular",
          "fontSize": 14,
          "lineHeight": 20,
        },
        Object {
          "color": "#0D111C",
        },
      ]
    }
  >
    Some very long text
  </Text>
</View>
`;
